crates/transaction-generator-lib/src/workflow_delegator.rs

```diff
@@ -0,0 +1,257 @@
+}
+
+impl StageTracking {
+    fn load_current_stage(&self) -> usize {
+        match self {
+            StageTracking::ExternallySet(stage) | StageTracking::WhenDone(stage) => {
+                stage.load(Ordering::Relaxed)
+            },
+        }
+    }
+}
+
+/// Generator allowing for multi-stage workflows.
+/// List of generators are passed:
+/// gen_0, gen_1, ... gen_n
+/// and on list of account pools, each representing accounts in between two stages:
+/// pool_0, pool_1, ... pool_n-1
+///
+/// pool_i is filled by gen_i, and consumed by gen_i+1, and so there is one less pools than generators.
+///
+/// We start with stage 0, which calls gen_0 pool_per_stage times, which populates pool_0 with accounts.
+///
+/// After that, in stage 1, we call gen_1, which consumes accounts from pool_0, and moves them to pool_1.
+/// We do this until pool_0 is empty.
+///
+/// We proceed, until in the last stage - stage n - calls gen_n, which consumes accounts from pool_n-1.
+///
+/// There are two modes on when to move to the next stage:
+/// - WhenDone means as soon as pool_i is empty, we move to stage i+1
+/// - ExternallySet means we wait for external signal to move to next stage, and we stop creating transactions
+///   until we receive it (or will move early if pool hasn't been consumed yet)
+///
+/// Use WorkflowTxnGeneratorCreator::create_workload to create this generator.
+struct WorkflowTxnGenerator {
+    stage: StageTracking,
+    generators: Vec<Box<dyn TransactionGenerator>>,
+    pool_per_stage: Vec<Arc<ObjectPool<LocalAccount>>>,
+    num_for_first_stage: usize,
+    // Internal counter, so multiple workers (WorkflowTxnGenerator) can coordinate how many times to execute the first stage
+    completed_for_first_stage: Arc<AtomicUsize>,
+}
+
+impl WorkflowTxnGenerator {
+    fn new(
+        stage: StageTracking,
+        generators: Vec<Box<dyn TransactionGenerator>>,
+        pool_per_stage: Vec<Arc<ObjectPool<LocalAccount>>>,
+        num_for_first_stage: usize,
+        completed_for_first_stage: Arc<AtomicUsize>,
+    ) -> Self {
+        Self {
+            stage,
+            generators,
+            pool_per_stage,
+            num_for_first_stage,
+            completed_for_first_stage,
+        }
+    }
+}
+
+impl TransactionGenerator for WorkflowTxnGenerator {
+    fn generate_transactions(
+        &mut self,
+        account: &LocalAccount,
+        mut num_to_create: usize,
+    ) -> Vec<SignedTransaction> {
+        assert_ne!(num_to_create, 0);
+        let mut stage = self.stage.load_current_stage();
+
+        match &self.stage {
+            StageTracking::WhenDone(stage_counter) => {
+                if stage == 0 {
+                    let prev = self
+                        .completed_for_first_stage
+                        .fetch_add(num_to_create, Ordering::Relaxed);
+                    num_to_create =
+                        cmp::min(num_to_create, self.num_for_first_stage.saturating_sub(prev));
```

Comment:
This is neat. But could you please explain how this actually works. That is each thread captures the intermediate value of 'completed_for_first_stage' in 'prev' and then sets the 'num_to_create' accordingly. Perhaps renaming prev to a more indicative name could help

Reply:
completed_for_first_stage.fetch_add(num_to_create) is atomic operation, and returns value before addition was applied. 

so everyone basically gets a chunk from the [0, +inf) array of size num_to_create

so if everyone calls that, some might get values below and some might get values above num_for_first_stage. So each thread should only do work for slot in the [0, num_for_first_stage)

makes sense?

Reply:
Sorry, I meant it could be great if you add a comment explaining this or change prev to more indicative name. 
I understand the logic